src/connection/websocket-handler.jl

```diff
@@ -0,0 +1,114 @@
+using HTTP.WebSockets: WebSocket, WebSocketError
+using HTTP.WebSockets: receive, isclosed
+using HTTP.WebSockets
+
+mutable struct WebSocketHandler
+    socket::Union{Nothing,WebSocket}
+    lock::ReentrantLock
+end
+
+WebSocketHandler(socket) = WebSocketHandler(socket, ReentrantLock())
+WebSocketHandler() = WebSocketHandler(nothing, ReentrantLock())
+
+function safe_read(websocket)
+    try
+        # readavailable is what HTTP overloaded for websockets
+        return receive(websocket)
+    catch e
+        if WebSockets.isok(e)
+            # it's ok :shrug:
+        elseif e isa Union{Base.IOError, EOFError}
+            @info("WS connection closed because of IO error")
+            return nothing
+        else
+            rethrow(e)
+        end
+    end
+end
+
+function safe_write(websocket, binary)
+    try
+        # send is what HTTP overloaded for writing to a websocket
+        send(websocket, binary)
+        return true
+    catch e
+        if WebSockets.isok(e) || e isa Union{Base.IOError,EOFError}
+            @warn "sending message to a closed websocket" maxlog = 1
+            # it's ok :shrug:
+            return nothing
+        else
+            rethrow(e)
+        end
+    end
+end
+
+function Base.isopen(ws::WebSocketHandler)
+    isnothing(ws.socket) && return false
+    # isclosed(ws.socket) returns readclosed && writeclosed
+    # but we consider it closed if either is closed?
+    if ws.socket.readclosed || ws.socket.writeclosed
+        return false
+    end
+    # So, it turns out, ws connection where the tab gets closed
+    # stay open indefinitely, but aren't writable anymore
+    # TODO, figure out how to check for that
+    return true
+end
+
+function Base.write(ws::WebSocketHandler, binary)
+    if isnothing(ws.socket)
+        error("socket closed or not opened yet")
+    end
+    lock(ws.lock) do
+        written = safe_write(ws.socket, binary)
+        if written != true
+            @debug "couldnt write, closing ws"
+            close(ws)
+        end
+    end
+end
+
+function Base.close(ws::WebSocketHandler)
+    isnothing(ws.socket) && return
+    try
+        socket = ws.socket
+        ws.socket = nothing
+        isclosed(socket) || close(socket)
+    catch e
+        if !WebSockets.isok(e)
+            @warn "error while closing websocket" exception=(e, Base.catch_backtrace())
+        end
+    end
+end
+
+
+"""
+    runs the main connection loop for the websocket
+"""
+function run_connection_loop(session::Session, handler::WebSocketHandler, websocket::WebSocket)
+    @debug("opening ws connection for session: $(session.id)")
+    handler.socket = websocket
+    while !isclosed(websocket)
```

Comment:
Should this be `isopen(handler)` ?

Reply:
This line is copied from https://github.com/SimonDanisch/Bonito.jl/blob/1964ec02972e5e31c496ad1ac8e20beab81bc26b/src/connection/websocket.jl#L95

The reference for the method is here https://juliaweb.github.io/HTTP.jl/stable/reference/#HTTP.WebSockets.isclosed

Do you mean Base.isopen? Is it the opposite?

Reply:
Okay done since it seems fine to change then. It should just mean we get rid of half-closed sockets quicker I suppose.

Reply:
Yeah that's what I hope :)  Although, the web does keep surprising me!